common/src/main/java/com/microsoft/identity/common/internal/telemetry/Telemetry.java

```diff
@@ -0,0 +1,283 @@
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_CORRELATION_ID;
+
+public class Telemetry {
+    private final static String TAG = Telemetry.class.getSimpleName();
+    private static volatile Telemetry singleton = null;
+    private List<ITelemetryObserver> mObservers;
+    private List<Map<String, String>> mTelemetryRawDataMap;
+    private TelemetryConfiguration mDefaultConfiguration;
+    private TelemetryContext mTelemetryContext;
+    private boolean mIsDebugging;
+
+    //if the telemetry configuration is not set
+    private final boolean mIsTelemetryEnabled;
+
+    private Telemetry(final Builder builder) {
+        if (builder == null) {
+            throw new IllegalArgumentException("Builder cannot be null");
+        }
+
+        if (builder.mTelemetryContext == null) {
+            throw new IllegalArgumentException("Telemetry context cannot be null");
+        }
+
+        if (builder.mDefaultConfiguration == null) {
+            //if the telemetry configuration is not set in the json, disable the telemetry.
+            mIsTelemetryEnabled = false;
+        } else {
+            mIsTelemetryEnabled = true;
+            mDefaultConfiguration = builder.mDefaultConfiguration;
+            mTelemetryContext = builder.mTelemetryContext;
+            mIsDebugging = builder.mIsDebugging;
+            mTelemetryRawDataMap = new LinkedList<>();
+            mObservers = new LinkedList<>();
+        }
+    }
+
+    /**
+     * Prepares instance using builder.
+     **/
+    private static Telemetry prepareInstance(Builder builder) {
+        if (singleton == null) {
+            synchronized (Telemetry.class) {
+                if (singleton == null) {
+                    singleton = new Telemetry(builder);
+                }
+            }
+        }
+        return singleton;
+    }
+
+    /**
+     * This is for getting instance of Telemetry
+     **/
+    public static Telemetry getInstance() {
+        return singleton;
+    }
+
+    public TelemetryContext getTelemetryContext() {
+        return mTelemetryContext;
+    }
+
+    public List<Map<String, String>> getRequestMap() {
+        return mTelemetryRawDataMap;
+    }
+
+    /**
+     * Register the observer to upload the telemetry data.
+     *
+     * @param observer ITelemetryReceiver
+     */
+    public synchronized void addObserver(final ITelemetryObserver observer) {
+        if (null == observer) {
+            throw new IllegalArgumentException("Receiver instance cannot be null");
+        }
+
+        // check to make sure we're not already dispatching elsewhere
+        if (null == mObservers) {
+            mObservers = new LinkedList<>();
+        }
+
+        mObservers.add(observer);
+    }
+
+    public synchronized void removeObserver(final ITelemetryObserver observer) {
+        if (null == observer || null == mObservers) {
+            return;
```

Comment:
nit: add logging point?

Reply:
I've talked with Shane before that one aim here is to make Telemetry calls as "quiet" as possible and not break any request because of the telemetry calls. We can follow-up tmr morning if this is still a concern.